grpc/interop/src/main/scala/io/buoyant/grpc/interop/Client.scala

```diff
@@ -0,0 +1,241 @@
+    val body = Buf.ByteArray.Owned(Array.fill(sz) { 0.toByte })
+    pb.Payload(body = Some(body))
+  }
+
+  private def unimplementedTest(name: String) =
+    Future.exception(new UnimplementedException(name))
+
+  class UnimplementedException(name: String) extends Exception(s"test not implemented: '${name}'")
+}
+
+class Client(
+  svc: pb.TestService
+) {
+  import Client.{mkPayload, unimplementedTest}
+
+  def emptyUnary(): Future[Unit] =
+    svc.emptyCall(pb.Empty()).flatMap {
+      case pb.Empty() => Future.Unit
+      case rsp => Future.exception(new IllegalArgumentException(s"unexpected response: $rsp"))
+    }
+
+  def largeUnary(reqSize: Int, rspSize: Int): Future[Unit] = {
+    val req = pb.SimpleRequest(
+      responseType = Some(pb.PayloadType.COMPRESSABLE), // cargocult
+      responseSize = Some(rspSize),
+      payload = Some(mkPayload(reqSize))
+    )
+    svc.unaryCall(req).flatMap {
+      case pb.SimpleResponse(Some(pb.Payload(_, Some(buf))), _, _) =>
+        if (buf.length == rspSize) Future.Unit
+        else Future.exception(new IllegalArgumentException(s"received ${buf.length}B, expected ${rspSize}B"))
+
+      case rsp => Future.exception(new IllegalArgumentException(s"unexpected response: $rsp"))
+    }
+  }
+
+  def clientStreaming(reqSizes: Seq[Int]): Future[Unit] = {
+    val reqs = Stream.mk[pb.StreamingInputCallRequest]
+
+    def sendReqs(szs0: Seq[Int]): Future[Unit] = szs0 match {
+      case Nil => reqs.close()
+      case Seq(sz, szs1@_*) =>
+        val req = pb.StreamingInputCallRequest(payload = Some(mkPayload(sz)))
+        reqs.send(req).before(sendReqs(szs1))
+    }
+
+    sendReqs(reqSizes).join(svc.streamingInputCall(reqs)).flatMap {
+      case (_, pb.StreamingInputCallResponse(Some(sz))) =>
+        val sum = reqSizes.sum
+        if (sz == sum) Future.Unit
+        else Future.exception(new IllegalArgumentException(s"received ${sz}B, expected ${sum}B"))
+      case (_, rsp) =>
+        Future.exception(new IllegalArgumentException(s"unexpected response: $rsp"))
+    }
+  }
+
+  def serverStreaming(rspSizes0: Seq[Int]): Future[Unit] = {
+    val req = pb.StreamingOutputCallRequest(
+      responseParameters = rspSizes0.map { sz => pb.ResponseParameters(size = Some(sz)) }
+    )
+    val rsps = svc.streamingOutputCall(req)
+
+    def read(rspSizes: Seq[Int]): Future[Unit] = rspSizes match {
+      case Nil => Future.Unit
+      case Seq(expected, rest@_*) =>
+        rsps.recv().transform {
+          case Throw(GrpcStatus.Ok(_)) => Future.Unit
+          case Throw(e) => Future.exception(e)
+          case Return(Stream.Releasable(rsp, release)) => rsp match {
+            case pb.StreamingOutputCallResponse(Some(pb.Payload(_, Some(buf)))) =>
+              val sz = buf.length
+              release().before {
+                if (sz != expected) {
+                  Future.exception(new IllegalArgumentException(s"recieved ${sz}B, expected ${expected}B"))
+                } else read(rest)
```

Comment:
it's tail recursive-ish. (this is a common idiom in finagle)

It can't actually be marked tail-recursive because we're doing `release().before(read(rest))` (which is equiv to `release().flatMap(_ => read(rest))`) -- since the recursive call is in the closure, the compiler can't know that it's in the tail position. Futures collapse chaining properly, though, and so this is all fine.